File: agent/linux/pkg/remediation/RemediationEngine.go
package remediation

import (
	"fmt"
	"log"
	"os"
	"syscall"

	"github.com/AntiVirus/agent/linux/pkg/quarantine"
)

type RemediationAction string

const (
	ActionKillProcess RemediationAction = "KillProcess"
	ActionQuarantine  RemediationAction = "Quarantine"
	ActionDelete      RemediationAction = "Delete"
	ActionAlert       RemediationAction = "Alert"
)

type RemediationRequest struct {
	Pid      uint32
	FilePath string
	Reason   string
	Actions  []RemediationAction
}

type RemediationResult struct {
	Success bool
	Actions []string
	Errors  []string
}

type RemediationEngine struct {
	quarantine *quarantine.QuarantineStore
}

func NewRemediationEngine(q *quarantine.QuarantineStore) *RemediationEngine {
	return &RemediationEngine{quarantine: q}
}

func (e *RemediationEngine) Remediate(req RemediationRequest) RemediationResult {
	result := RemediationResult{Success: true}

	for _, action := range req.Actions {
		switch action {
		case ActionKillProcess:
			if err := e.killProcess(req.Pid); err != nil {
				msg := fmt.Sprintf("KillProcess PID %d failed: %v", req.Pid, err)
				log.Println("[RemediationEngine]", msg)
				result.Errors = append(result.Errors, msg)
				result.Success = false
			} else {
				result.Actions = append(result.Actions, fmt.Sprintf("Killed PID %d", req.Pid))
			}

		case ActionQuarantine:
			if req.FilePath == "" {
				continue
			}
			entry, err := e.quarantine.Quarantine(req.FilePath, req.Reason)
			if err != nil {
				msg := fmt.Sprintf("Quarantine %s failed: %v", req.FilePath, err)
				log.Println("[RemediationEngine]", msg)
				result.Errors = append(result.Errors, msg)
				result.Success = false
			} else {
				result.Actions = append(result.Actions, fmt.Sprintf("Quarantined %s -> %s", req.FilePath, entry.QuarantinePath))
			}

		case ActionDelete:
			if req.FilePath == "" {
				continue
			}
			if err := os.Remove(req.FilePath); err != nil {
				msg := fmt.Sprintf("Delete %s failed: %v", req.FilePath, err)
				log.Println("[RemediationEngine]", msg)
				result.Errors = append(result.Errors, msg)
				result.Success = false
			} else {
				result.Actions = append(result.Actions, fmt.Sprintf("Deleted %s", req.FilePath))
			}

		case ActionAlert:
			log.Printf("[RemediationEngine][ALERT] %s | File: %s | PID: %d", req.Reason, req.FilePath, req.Pid)
			result.Actions = append(result.Actions, "Alert logged")
		}
	}

	return result
}

func (e *RemediationEngine) killProcess(pid uint32) error {
	proc, err := os.FindProcess(int(pid))
	if err != nil {
		return err
	}
	if err := proc.Signal(syscall.SIGKILL); err != nil {
		return err
	}
	log.Printf("[RemediationEngine] Sent SIGKILL to PID %d", pid)
	return nil
}

File: agent/macos/Sources/FenrirAgent/RemediationEngine.swift
import Foundation

enum RemediationAction {
    case killProcess
    case quarantineFile
    case deleteFile
    case alertOnly
}

struct RemediationRequest {
    let pid: pid_t
    let filePath: String?
    let reason: String
    let actions: [RemediationAction]
}

struct RemediationResult {
    var success: Bool = true
    var actionsPerformed: [String] = []
    var errors: [String] = []
}

class RemediationEngine {
    private let quarantineStore: QuarantineStore

    init(quarantineStore: QuarantineStore) {
        self.quarantineStore = quarantineStore
    }

    @discardableResult
    func remediate(_ request: RemediationRequest) -> RemediationResult {
        var result = RemediationResult()

        for action in request.actions {
            switch action {
            case .killProcess:
                if kill(request.pid, SIGKILL) == 0 {
                    let msg = "Sent SIGKILL to PID \(request.pid)"
                    print("[RemediationEngine] \(msg)")
                    result.actionsPerformed.append(msg)
                } else {
                    let err = "Failed to kill PID \(request.pid): \(String(cString: strerror(errno)))"
                    print("[RemediationEngine] \(err)")
                    result.errors.append(err)
                    result.success = false
                }

            case .quarantineFile:
                guard let path = request.filePath else { continue }
                let ok = quarantineStore.quarantine(filePath: path, reason: request.reason)
                if ok {
                    result.actionsPerformed.append("Quarantined \(path)")
                } else {
                    let err = "Quarantine failed for \(path)"
                    result.errors.append(err)
                    result.success = false
                }

            case .deleteFile:
                guard let path = request.filePath else { continue }
                do {
                    try FileManager.default.removeItem(atPath: path)
                    let msg = "Deleted \(path)"
                    print("[RemediationEngine] \(msg)")
                    result.actionsPerformed.append(msg)
                } catch {
                    let err = "Delete failed for \(path): \(error)"
                    result.errors.append(err)
                    result.success = false
                }

            case .alertOnly:
                let msg = "[ALERT] \(request.reason) | PID: \(request.pid) | File: \(request.filePath ?? "N/A")"
                print("[RemediationEngine] \(msg)")
                result.actionsPerformed.append("Alert logged")
            }
        }

        return result
    }
}
